File: plugins/TitleManager/src/main/java/io/puharesource/mc/titlemanager/commands/TMCommand.java
package io.puharesource.mc.titlemanager.commands;

import io.puharesource.mc.titlemanager.TitleManager;
import io.puharesource.mc.titlemanager.backend.language.Messages;
import lombok.val;
import org.bukkit.ChatColor;
import org.bukkit.command.*;

import java.util.*;

public final class TMCommand implements CommandExecutor, TabCompleter {

    private Map<String, TMSubCommand> commands = new HashMap<>();

    public TMCommand() {
        PluginCommand cmd = TitleManager.getInstance().getCommand("tm");
        cmd.setExecutor(this);
        cmd.setTabCompleter(this);
    }

    public void addSubCommand(TMSubCommand cmd) {
        commands.put(cmd.getAlias().toUpperCase(), cmd);
        for (String alias : cmd.getAliases())
            commands.put(alias.toUpperCase(), cmd);
    }

    private void syntaxError(CommandSender sender) {
        sender.sendMessage(ChatColor.RED + Messages.COMMAND_WRONG_USAGE.getMessage());
        List<String> aliases = new ArrayList<>();
        for (TMSubCommand cmd : commands.values()) {
            if (aliases.contains(cmd.getAlias().toUpperCase())) continue;
            sender.sendMessage(String.format(ChatColor.RED + Messages.COMMAND_WRONG_USAGE_TRAIL.getMessage(), "/tm " + cmd.getAlias(), cmd.getUsage() + ChatColor.GRAY + (cmd.getUsage().isEmpty() ? "- " : " - ") + cmd.getDescription()));
            aliases.add(cmd.getAlias().toUpperCase());
        }
    }

    @Override
    public boolean onCommand(CommandSender sender, Command cmd, String label, String[] args) {
        if (!cmd.getName().equalsIgnoreCase("tm")) return false;
        if (args.length >= 1) {
            TMSubCommand subCommand = commands.get(args[0].toUpperCase());
            if (subCommand != null)
                if (sender.hasPermission(subCommand.getNode())) {
                    String[] subArgs = Arrays.copyOfRange(args, 1, args.length);
                    val parameters = getParameters(subCommand, subArgs);
                    subCommand.onCommand(sender, Arrays.copyOfRange(subArgs, parameters.getParams().size(), subArgs.length), parameters);
                } else sender.sendMessage(ChatColor.RED + Messages.COMMAND_NO_PERMISSION.getMessage());
            else syntaxError(sender);
        } else syntaxError(sender);
        return true;
    }

    @Override
    public List<String> onTabComplete(CommandSender sender, Command cmd, String label, String[] args) {
        List<String> possibilities = new ArrayList<>();

        if (args.length == 1)
            for (String sub : commands.keySet())
                if (sub.toLowerCase().startsWith(args[0].toLowerCase()))
                    possibilities.add(sub.toLowerCase());

        return possibilities.size() <= 0 ? null : possibilities;
    }

    private CommandParameters getParameters(final TMSubCommand cmd, final String[] args) {
        final Collection<String> supportedParameters = cmd.getSupportedParameters();
        final Map<String, CommandParameter> parameters = new HashMap<>();

        for (String arg : args) {
            if (!arg.startsWith("-")) break;

            char[] chars = arg.toCharArray();
            String fullParameter = "";
            for (int i = 1; chars.length > i; i++) {
                fullParameter += chars[i];
            }

            if (fullParameter.contains("=")) {
                String[] paramValues = fullParameter.split("=", 2);
                String param = paramValues[0].toUpperCase();

                if (supportedParameters.contains(param)) {
                    parameters.put(param, new CommandParameter(param, paramValues[1]));
                } else break;
            } else {
                String param = fullParameter.toUpperCase();
                if (supportedParameters.contains(param)) {
                    parameters.put(param, new CommandParameter(param, null));
                } else break;
            }
        }

        return new CommandParameters(parameters);
    }
}

File: src/main/kotlin/io/puharesource/mc/titlemanager/commands/TMCommand.kt
package io.puharesource.mc.titlemanager.commands

import io.puharesource.mc.titlemanager.extensions.sendConfigMessage
import org.bukkit.ChatColor
import org.bukkit.command.Command
import org.bukkit.command.CommandExecutor
import org.bukkit.command.CommandSender
import org.bukkit.command.TabCompleter
import java.util.TreeMap
import java.util.TreeSet

object TMCommand : CommandExecutor, TabCompleter {
    private val subCommands : MutableMap<String, TMSubCommand> = TreeMap(String.CASE_INSENSITIVE_ORDER)
    private val parameterPattern = """(?i)^[-](silent|world|fadein|stay|fadeout|radius)$""".toRegex()
    private val parameterPatternWithValue = """(?i)^[-](silent|world|fadein|stay|fadeout|radius)[=]([^ ]+)$""".toRegex()

    init {
        addSubCommand(CommandBroadcast)
        addSubCommand(CommandABroadcast)
        addSubCommand(CommandMessage)
        addSubCommand(CommandAMessage)
        addSubCommand(CommandScoreboard)
        addSubCommand(CommandAnimations)
        addSubCommand(CommandScripts)
        addSubCommand(CommandReload)
        addSubCommand(CommandVersion)
    }

    fun addSubCommand(cmd: TMSubCommand) {
        subCommands[cmd.name] = cmd
        cmd.aliases.forEach { subCommands[it] = cmd }
    }

    /**
     * Sends a list of commands, usages and descriptions to a command sender.

     * @param sender The command sender to send the list to.
     * *
     * @param primaryColor The primary color of the list.
     * *
     * @param secondaryColor The secondary color of the list.
     */
    private fun commandList(sender: CommandSender, primaryColor: ChatColor, secondaryColor: ChatColor) {
        // A set of commands aliases that have already been formatted.
        val alreadyShown = TreeSet(String.CASE_INSENSITIVE_ORDER)

        // Iterate over all aliases and send the formatted message.
        for (cmd in subCommands.values) {
            // Continue if the command has already been shown.
            if (alreadyShown.contains(cmd.name)) continue

            // Continue if the player doesn't have permission to use the command.
            if (cmd.permission.isNotBlank() && !sender.hasPermission(cmd.permission)) continue

            // Add the command's name to the list of shown commands, to avoid showing the same command multiple times.
            alreadyShown.add(cmd.name)

            // Send the formatted message.
            sender.sendMessage(formatCommandListString(cmd, primaryColor, secondaryColor))
        }
    }

    /**
     * Gets a formatted command, usage and description string, from the specified command.
     *
     * @param subCommand If the command to be described is a sub command, then specify an empty optional.
     * @param primaryColor The primary color to be used.
     * @param secondaryColor The secondary color to be used.
     * @return The line
     */
    private fun formatCommandListString(subCommand: TMSubCommand, primaryColor: ChatColor, secondaryColor: ChatColor): String {
        // Create the string builder, to build the message.
        val sb = StringBuilder("$primaryColor").append("    /tm")

        sb.append(" ").append(subCommand.name.toLowerCase())

        // If a usage is specified, add the usage to the message.
        if (subCommand.usage.isNotBlank()) {
            sb.append(" ").append(subCommand.usage)
        }

        // If a description is specified, add the description to the message.
        if (subCommand.description.isNotBlank()) {
            sb.append("$secondaryColor").append(" - ").append(subCommand.description)
        }

        return sb.toString()
    }

    fun syntaxError(sender: CommandSender) {
        sender.sendMessage("${ChatColor.RED}Wrong usage! Correct usage" + if (subCommands.size > 1) "s:" else ":")
        commandList(sender, ChatColor.RED, ChatColor.GRAY)
    }

    fun helpMessage(sender: CommandSender) {
        sender.sendMessage(String.format("${ChatColor.YELLOW}Command \"/tm\" help:"))
        commandList(sender, ChatColor.YELLOW, ChatColor.GRAY)
    }

    override fun onCommand(sender: CommandSender, cmd: Command, label: String, args: Array<out String>): Boolean {
        if (!cmd.name.equals("tm", ignoreCase = true)) return false

        if (args.isEmpty()) {
            syntaxError(sender)
            return true
        }

        if (args.first().equals("help", ignoreCase = true) || args.first().equals("?", ignoreCase = true)) {
            helpMessage(sender)
            return true
        }

        val subCommand = subCommands[args.first()]

        if (subCommand == null) {
            syntaxError(sender)
            return true
        }

        if (subCommand.permission.isNotBlank() && !sender.hasPermission(subCommand.permission)) {
            sender.sendConfigMessage("no-permission")
            return true
        }

        if (!subCommand.allowedSender.isAllowed(sender)) {
            sender.sendMessage("${ChatColor.RED}${subCommand.allowedSender.disallowMessage}")
            return true
        }

        val parameters : MutableMap<String, CommandParameter> = TreeMap(String.CASE_INSENSITIVE_ORDER)

        var i = 0
        for (arg in args) {
            if (i == 0) {
                i++
                continue
            }

            if (arg.matches(parameterPattern)) {
                val parameter = parameterPattern.matchEntire(arg)!!.groups[1]!!.value

                parameters[parameter] = CommandParameter(parameter)
            } else if (arg.matches(parameterPatternWithValue)) {
                val result = parameterPatternWithValue.matchEntire(arg)!!.groups
                val parameter = result[1]!!.value
                val value = result[2]!!.value

                parameters[parameter] = CommandParameter(parameter, value)
            } else {
                break
            }

            i++
        }

        subCommand.runCommand(sender, args.copyOfRange(i, args.size), parameters)
        return true
    }

    override fun onTabComplete(sender: CommandSender, command: Command, alias: String, args: Array<out String>): MutableList<String>? {
        if (args.isNotEmpty()) {
            return subCommands.keys
                    .asSequence()
                    .filter { it.startsWith(args[args.size - 1], ignoreCase = true) }
                    .toMutableList()
        }

        return null
    }
}